osquery/sql/sqlite_collation.cpp

```diff
@@ -0,0 +1,110 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include <regex>
+
+#include <osquery/core/flags.h>
+
+#include <sqlite3.h>
+
+namespace osquery {
+/**
+ * @brief Split version string into numeric and alphabetic character segments.
+ */
+static std::vector<std::vector<std::string>> versionSplit(
+    const std::string& version) {
+  std::vector<std::vector<std::string>> result;
+  std::regex ver_pattern("(\\d+|[a-zA-Z]+)");
+  std::sregex_token_iterator iter_end;
+  std::istringstream iss(version);
+  std::string segment;
+
+  // Split version into segments on periods.
+  while (getline(iss, segment, '.')) {
+    std::vector<std::string> seg_pieces;
+    std::sregex_token_iterator iter_begin(
+        segment.begin(), segment.end(), ver_pattern);
+
+    // Split segment into pieces of consecutive alphabetic and numeric
+    // characters.
+    while (iter_begin != iter_end) {
+      seg_pieces.push_back(*iter_begin++);
+    }
+
+    result.push_back(seg_pieces);
+  }
+
+  return result;
+}
+
+/**
+ * @brief Collate version strings. (Only compares alphanumeric characters.)
+ */
+static int versionCollate(
+    void* notUsed, int nKey1, const void* pKey1, int nKey2, const void* pKey2) {
+  const std::string lver(static_cast<const char*>(pKey1), nKey1);
+  const std::string rver(static_cast<const char*>(pKey2), nKey2);
+  (void)notUsed;
+
+  // Early return if versions are equal.
+  if (lver == rver) {
+    return 0;
+  }
+
+  // Get version segments with their nested pieces.
+  auto lver_vs = versionSplit(lver);
+  auto rver_vs = versionSplit(rver);
```

Comment:
I'm slightly concerned about the overhead of split if this is used on a large table (like dpkg or rpm versions). And, I don't think you need it -- you ought be able to walk the strings L to R and compare ASCII values.

Reply:
And/or use `std::string_view` whenever possible, like splitting and other operations that do not really require a copy of the data.

Reply:
Yeah, I'm not sure why I committed to this fully, but the new version is hopefully better. Good to know about the `std::string_view`.